Mathlib/GroupTheory/Perm/Cycle/Factors.lean

```diff
@@ -42,19 +35,101 @@ section CycleOf
-#align equiv.perm.cycle_of Equiv.Perm.cycleOf
 
-theorem cycleOf_apply (f : Perm α) (x y : α) :
-    cycleOf f x y = if SameCycle f x y then f y else y := by
-  dsimp only [cycleOf]
+private theorem cycleOfGDR_apply (f : Perm α) (x y : α) :
+    cycleOfGDR f x y = if SameCycle f x y then f y else y := by
+  dsimp only [cycleOfGDR]
   split_ifs with h
   · apply ofSubtype_apply_of_mem
     exact h
   · apply ofSubtype_apply_of_not_mem
     exact h
+
+private theorem cycleOfGDR_eq_one_iff (f : Perm α) : cycleOfGDR f x = 1 ↔ f x = x := by
+  simp_rw [ext_iff, cycleOfGDR_apply, one_apply]
+  refine ⟨fun h => (if_pos (SameCycle.refl f x)).symm.trans (h x), fun h y => ?_⟩
+  by_cases hy : f y = y
+  · rw [hy, ite_self]
+  · exact if_neg (mt SameCycle.apply_eq_self_iff (by tauto))
+
+private theorem mem_support_cycleOfGDR_iff :
+    y ∈ support (f.cycleOfGDR x) ↔ SameCycle f x y ∧ x ∈ support f := by
+  by_cases hx : f x = x
+  · rw [(cycleOfGDR_eq_one_iff _).mpr hx]
+    simp [hx]
+  · rw [mem_support, cycleOfGDR_apply]
+    split_ifs with hy
+    · simp only [hx, hy, iff_true_iff, Ne, not_false_iff, and_self_iff, mem_support]
+      rcases hy with ⟨k, rfl⟩
+      rw [← not_mem_support]
+      simpa using hx
+    · simpa [hx] using hy
+
+private theorem mem_support_cycleOfGDR_iff' (hx : f x ≠ x) :
+    y ∈ support (f.cycleOfGDR x) ↔ SameCycle f x y := by
+  rw [mem_support_cycleOfGDR_iff, and_iff_left (mem_support.2 hx)]
+
+private theorem isCycleOn_support_cycleOfGDR (f : Perm α) (x : α) :
+    f.IsCycleOn (f.cycleOfGDR x).support :=
+  ⟨f.bijOn <| by
+    refine fun _ =>
+      ⟨fun h ↦ mem_support_cycleOfGDR_iff.2 ?_, fun h ↦ mem_support_cycleOfGDR_iff.2 ?_⟩
+    · exact ⟨sameCycle_apply_right.1 (mem_support_cycleOfGDR_iff.1 h).1,
+      (mem_support_cycleOfGDR_iff.1 h).2⟩
+    · exact ⟨sameCycle_apply_right.2 (mem_support_cycleOfGDR_iff.1 h).1,
+      (mem_support_cycleOfGDR_iff.1 h).2⟩
+    , fun a ha b hb =>
+      by
+        rw [mem_coe, mem_support_cycleOfGDR_iff] at ha hb
+        exact ha.1.symm.trans hb.1⟩
+
+private theorem SameCycle.exists_lt_card_support_cycleOfGDR_pow_eq_of_mem_support
+    (h : SameCycle f x y) (hx : x ∈ f.support) :
+    ∃ i < (f.cycleOfGDR x).support.card, (f ^ i) x = y := by
+  rw [mem_support] at hx
+  exact Equiv.Perm.IsCycleOn.exists_pow_eq (b := y) (f.isCycleOn_support_cycleOfGDR x)
+    (by rw [mem_support_cycleOfGDR_iff' hx]) (by rwa [mem_support_cycleOfGDR_iff' hx])
+
+end GeneralizedDecRel
+
+instance instDecidableRelSameCycle (f : Perm α) : DecidableRel (SameCycle f) := fun x y =>
+  decidable_of_iff (y ∈ List.iterate f x (Fintype.card α)) <| by
+    simp only [List.mem_iterate, iterate_eq_pow, eq_comm (a := y)]
+    constructor
+    · rintro ⟨n, _, hn⟩
+      exact ⟨n, hn⟩
+    · intro hxy
+      by_cases hx : x ∈ f.support
+      case pos =>
+        classical
+        rcases hxy.exists_lt_card_support_cycleOfGDR_pow_eq_of_mem_support hx with ⟨i, hixy, hi⟩
```

Comment:
This pattern seems quite strange to me; is it one we use elsewhere in mathlib?

Reply:
As far as I know, there are none.

Reply:
Edited above with a comment suggestion